components/script/dom/subtlecrypto.rs

```diff
@@ -860,6 +864,268 @@ impl SubtleCryptoMethods for SubtleCrypto {
+                    }
+                    ALG_AES_CBC => {
+                        let KeyWrapAlgorithm::AesCbc(params) = normalized_algorithm else {
+                            promise.reject_error(Error::Data);
+                            return;
+                        };
+                        subtle.decrypt_aes_cbc(
+                            &params, &unwrapping_key, &wrapped_key_bytes, cx, array_buffer_ptr.handle_mut()
+                        )
+                    },
+                    ALG_AES_CTR => {
+                        let KeyWrapAlgorithm::AesCtr(params) = normalized_algorithm else {
+                            promise.reject_error(Error::Data);
+                            return;
+                        };
+                        subtle.encrypt_decrypt_aes_ctr(
+                            &params, &unwrapping_key, &wrapped_key_bytes, cx, array_buffer_ptr.handle_mut()
+                        )
+                    },
+                    _ => Err(Error::NotSupported),
+                };
+
+                let bytes = match result {
+                    Ok(bytes) => bytes,
+                    Err(e) => {
+                        promise.reject_error(e);
+                        return;
+                    },
+                };
+
+                let import_key_bytes = match format {
+                    KeyFormat::Raw | KeyFormat::Spki | KeyFormat::Pkcs8 => &bytes,
+                    KeyFormat::Jwk => {
+                        let jwk_string = String::from_utf8_lossy(&bytes.to_vec()).to_string();
+                        let Ok(value) = serde_json::from_str(&jwk_string) else {
+                            promise.reject_error(Error::Syntax);
+                            return;
+                        };
+                        let serde_json::Value::Object(obj) = value else {
+                            promise.reject_error(Error::Syntax);
+                            return;
+                        };
+                        let Some(ext_string) = obj.get("ext") else {
+                            promise.reject_error(Error::Syntax);
+                            return;
+                        };
+                        let serde_json::Value::Bool(ext) = ext_string else {
+                            promise.reject_error(Error::Syntax);
+                            return;
+                        };
+                        if !ext && extractable {
+                            promise.reject_error(Error::Data);
+                            return;
+                        }
+                        let Some(key_string) = obj.get("k") else {
+                            promise.reject_error(Error::Syntax);
+                            return;
+                        };
+                        let serde_json::Value::String(k_string) = key_string else {
+                            promise.reject_error(Error::Syntax);
+                            return;
+                        };
```

Comment:
A few questions about this block:
* Given that all the values here are required, is there some reason this can't use a struct with `#[derive(Deserialize)]`? Doing deserialization like this for the rsa-family of algorithms (which have many more fields) is going to be hurt.

* I believe this is implementing [parse a JWK](https://w3c.github.io/webcrypto/#concept-parse-a-jwk). The spec says:
  > If the [kty](https://w3c.github.io/webcrypto/#dom-jsonwebkey-kty) field of key is not defined, then [throw](https://webidl.spec.whatwg.org/#dfn-throw) a [DataError](https://webidl.spec.whatwg.org/#dataerror).

  While `unwrapKey` does not need the field, we still need to throw a `DataError` if it's not defined.

* I can't find the place in the spec that says we're supposed to do
   ```rust
    if !ext && extractable {
        promise.reject_error(Error::Data);
        return;
    }
    ```
   Am I overlooking something?
* I haven't looked at web platform tests for this part, but since the [spec](https://w3c.github.io/webcrypto/#concept-parse-a-jwk) says
  > Let key be the result of converting result to the IDL dictionary type of [JsonWebKey](https://w3c.github.io/webcrypto/#dfn-JsonWebKey).
  
  we should return `Error::Type` instead of `Error::Syntax` no?

Reply:
for 3), it's step 8 of importKey for AES-KW:
> If the [ext](https://w3c.github.io/webcrypto/#dom-jsonwebkey-ext) field of jwk is present and has the value false and extractable is true, then [throw](https://webidl.spec.whatwg.org/#dfn-throw) a [DataError](https://webidl.spec.whatwg.org/#dataerror).

No argument on your other points, I'll make those adjustments

Reply:
Thanks for the clarification